lxd/util/resources.go

```diff
@@ -0,0 +1,337 @@
+			if err != nil {
+				return nil, err
+			}
+
+			t.socketID = uint64(sockID)
+
+			path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_cur_freq", t.ID)
+			freq, err := parseNumberFromFile(path)
+			if err != nil {
+				return nil, err
+			}
+
+			t.frequency = uint64(freq / 1000)
+
+			path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_max_freq", t.ID)
+			freq, err = parseNumberFromFile(path)
+			if err != nil {
+				return nil, err
+			}
+
+			t.frequencyTurbo = uint64(freq / 1000)
+
+			threads = append(threads, *t)
+		} else if strings.HasPrefix(line, "vendor_id") {
+			i := strings.Index(line, ":")
+			if i < 0 {
+				return nil, err
+			}
+			i++
+
+			line = line[i:]
+			line = strings.TrimSpace(line)
+
+			if t != nil {
+				threads[len(threads)-1].name = line
+			}
+		} else if strings.HasPrefix(line, "model name") {
+			i := strings.Index(line, ":")
+			if i < 0 {
+				return nil, err
+			}
+			i++
+
+			line = line[i:]
+			line = strings.TrimSpace(line)
+
+			if t != nil {
+				threads[len(threads)-1].vendor = line
+			}
+		}
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
+	}
+
+	return threads, err
+}
+
+func parseSysDevSystemCpu() ([]thread, error) {
+	ents, err := ioutil.ReadDir("/sys/devices/system/cpu/")
+	if err != nil {
+		return nil, err
+	}
+
+	threads := []thread{}
+	for _, ent := range ents {
+		entName := ent.Name()
+		if !strings.HasPrefix(entName, "cpu") {
+			continue
+		}
+
+		entName = entName[len("cpu"):]
+		idx, err := strconv.Atoi(entName)
+		if err != nil {
+			continue
+		}
+
+		t := thread{}
+		t.ID = uint64(idx)
+		path := fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/core_id", t.ID)
+		coreID, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.coreID = uint64(coreID)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/topology/physical_package_id", t.ID)
+		sockID, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.socketID = uint64(sockID)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/scaling_cur_freq", t.ID)
+		freq, err := parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.frequency = uint64(freq / 1000)
+
+		path = fmt.Sprintf("/sys/devices/system/cpu/cpu%d/cpufreq/cpuinfo_max_freq", t.ID)
+		freq, err = parseNumberFromFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		t.frequencyTurbo = uint64(freq / 1000)
+
+		threads = append(threads, t)
+	}
+
+	if len(threads) == 0 {
+		return nil, db.NoSuchObjectError
+	}
+
+	return threads, err
+}
+
+func getThreads() ([]thread, error) {
+	threads, err := parseCpuinfo()
+	if err == nil {
+		return threads, nil
+	}
+
+	threads, err = parseSysDevSystemCpu()
+	if err != nil {
+		return nil, err
+	}
+
+	return threads, nil
+}
+
+func CpuResource() (api.ResourcesCPU, error) {
+	c := api.ResourcesCPU{}
+
+	threads, err := getThreads()
+	if err != nil {
+		return c, err
```

Comment:
Recently I've been trying to stay away from returning such empty structs as there's the risk that when adding another error check later, it won't be empty anymore.

How about changing this function to return a pointer and return nil on error?

Reply:
Agreed. It just seemed to be our previous pattern. I'd prefer pointers as well!

Reply:
Yeah, the lxd codebase is a bit inconsistent in that regard, the client/ package is much cleaner.

That is, it always returns pointers and takes structs as arguments when required or as pointers when optional, makes checking for nil much easier.

Reply:
done